### DISCUSSION TITLE
Exclude Camera Class From GigeVision Library

### DISCUSSION THREAD

**Discussion by FaisalFagihi** [Stira-sa/GigeVision#13] [General]
GigeVision library is about communicating with any camera that use GIGE protocols (GVCP, GVSP), it should not have anything related to camera functionalities.

Camera implementation should be done from the developer side, so it can be utilized to the cameras that he/she has.

Kindly check the following branch: 
[RefactorGenIcamAndGigeVisionCore](https://github.com/Stira-sa/GigeVision/tree/RefactorGenIcamAndGigeVisionCore)

**Comment by Touseefelahi** [comment]
The library is for GigeVision protocol based cameras, and it is generic

The basic camera implementation is not being forced to the developer, the developer can have his own implementation as desire

**Reply by FaisalFagihi** [reply]
Actually the current StreamReceiver(GVSP) depends on the current camera so the developer cannot use the GVSP without using the current Camera.

> public StreamReceiver(Camera camera)
>         {
>             Camera = camera;
>             GvspInfo = new GvspInfo();
>         }

**Reply by Touseefelahi** [reply]
Yes, then the StreamReceiver can be changed to have a generic implementation that will take the require inputs like IP and Port as string and int

**Reply by FaisalFagihi** [reply]
yes that's what is already been done in the Camera.Wpf project and will do the same for other protocols.

 sorry for talking about other projects in GigeVision repository but this the point to exclude something doesn't totally fit to it and place in more suitable, reusable, scalable  and testable place.

**Comment by Touseefelahi** [comment]
@FaisalFagihi I checked the branch and the implementation is not generic. It specifies that the camera is Daylight for some reason. This part is not supposed to be the part of the library, keeping the camera class general should be the way to go. DL/IR level differentiation should be done in the application layer.

**Reply by FaisalFagihi** [reply]
the goal here is that the application layer will deal with the ICamera only regardless of the protocol. 

the Camera.Wpf project will have multiple of profiles for GigeDLCamera, GigeIRCamera, RtspOnvifCamera,  and other will come.

the big advantage here is we don't have to test the camera functions multiple of times cause they will have the same  interface. 

the profiles could have many protocol as they need.

**Reply by FaisalFagihi** [reply]
not to mention that they all will have the same exact view and control.

**Reply by Touseefelahi** [reply]
The GigeVision library is meant to be used for Gige protocol based camera as mentioned before. So, I don't see any point to having RTSP/Onvif merged with the library project for some reason. 

The Camera.Wpf Implementation is supposed to be in the application layer and the Developer must implement it according to the application needs. 

To have different "Profiles" like DL and IR is totally according to the XML of the camera for the Library, I don't see any good reason to have these separate implementations of these in the library.

And for the testing, we want to do thorough testing for GVCP, GVSP, and StreamReceiver Class. The Camera Class utilizes the functionalities of the previously mentioned three classes and thus doesn't require thorough testing, simple StreamStart/Stop and SetResolution/Offset testing would be enough

**Reply by FaisalFagihi** [reply]
there is some misunderstanding.. allow me to clarify:
* Camera.Wpf is not part of GigeVision library at all it is there in the branch ([RefactorGenIcamAndGigeVisionCore](https://github.com/Stira-sa/GigeVision/tree/RefactorGenIcamAndGigeVisionCore)) just to demonstrate the structure. it totally different project and yes it will be used in the application layer.

the purpose of the Camera.Wpf is to handle variety of cameras protocols with one general interface.

**Reply by Touseefelahi** [reply]
Where does this Camera.Wpf belongs then? If not in the current GigeVision library then in some other library?

**Reply by FaisalFagihi** [reply]
yes we can name it as Stira.Camera

**Reply by Touseefelahi** [reply]
For anything to be the camera I guess it must have the following:
Image Resolution
Depth
No. Of Channels
Raw/Compressed
Compression Type for Compressed
Bayer Pattern for Raw Colored
and so on...

I would not like to go to tackle all this up to this level for a Graphics Side thing, that I believe is supposed to be implemented on Application side. 

I see there is advantage of making GigeVision.Camera.Wpf that is a graphics element of gige based camera only (Not having any other type as Onvif/RTSP)

**Reply by FaisalFagihi** [reply]
Good so we agreed on creating a new package dedicated for GigeVision camera implementation and to deprecate Camera and StreamReceiver classes from the core.

**Reply by Touseefelahi** [reply]
There is no harm of keeping the camera class inside the GigeVision as a core component without graphics.

If we want to have a graphics element as a GigeCamera viewer then it can be done in a separate project/package

Stream Receiver is also supposed to be the part of the GigeVision library

Only one thing can be done to separate the direct dependency of Camera for StreamReceiver class. I see they are tightly bound together because of some event handlers. If we manage these in an elegant way that would be better in the future, it is going to be time taking because of the criticality of the component.
